app/services/experimentation/data_export.rb

```diff
@@ -0,0 +1,214 @@
+        appointments_deficit.times do
+          patient_data[:appointments] << Array.new(10, nil)
+        end
+      end
+    end
+
+    def process_past_visits(patient)
+      end_date = notification_start_date - 1.day
+      start_date = end_date - 1.year
+      encounter_dates = encounter_dates(patient, start_date, end_date)
+      @max_past_visits = encounter_dates.count if encounter_dates.count > max_past_visits
+      encounter_dates
+    end
+
+    def pad_past_visits
+      aggregate.each do |patient_data|
+        past_visits_deficit = @max_past_visits - patient_data[:encounters].count
+        past_visits_deficit.times { patient_data[:encounters] << nil }
+      end
+    end
+
+    def encounter_dates(patient, start_date, end_date)
+      date_range = (start_date.beginning_of_day..end_date.end_of_day)
+      bps = patient.blood_pressures.where(device_created_at: date_range).pluck(:device_created_at).map(&:to_date)
+      bss = patient.blood_sugars.where(device_created_at: date_range).pluck(:device_created_at).map(&:to_date)
+      pds = patient.prescription_drugs.where(device_created_at: date_range).pluck(:device_created_at).map(&:to_date)
+      (bps + bss + pds).uniq.sort
+    end
+
+    def encounter_by_date(patient, date)
+      date_range = (date.beginning_of_day..date.end_of_day)
+      bp = patient.blood_pressures.find_by(device_created_at: date_range)
+      return bp if bp
+      bs = patient.blood_sugars.find_by(device_created_at: date_range)
+      return bs if bs
+      pd = patient.prescription_drugs.find_by(device_created_at: date_range)
+      return pd if pd
+      # this shouldn't happen. consider raising error
+    end
+
+    def headers
+      keys = aggregate.first.keys
+      keys.each_with_object([]) do |key, collection|
+        case key
+        when :communications
+          (1..@max_communications).step do |i|
+            collection << "Message #{i} Type"
+            collection << "Message #{i} Sent"
+            collection << "Message #{i} Status"
+            collection << "Message #{i} Text Identifier"
+          end
+        when :appointments
+          (1..@max_appointments).step do |i|
+            collection << "Appointment #{i} Creation Date"
+            collection << "Appointment #{i} Date"
+            collection << "Followup #{i} Date"
+            collection << "Days To Visit #{i}"
+            collection << "BP Recorded At Visit #{i}"
+            collection << "Patient Visited Facility #{i}"
+            collection << "Followup #{i} Facility Type"
+            collection << "Followup #{i} Facility State"
+            collection << "Followup #{i} Facility District"
+            collection << "Followup #{i} Facility Block"
+          end
+        when :encounters
+          (1..@max_past_visits).step do |i|
+            collection << "Encounter #{i} Date"
+          end
+        else
+          collection << key.to_s.split("_").each(&:capitalize!).join(" ")
+        end
+      end
+    end
```

Comment:
This is, uh, confusing 😁 ...I guess this is handling the case where there are 1 to many amounts of communications and appointments? This is really a case where we **will** want to have hashes, and then the CSV writing can be a bit smarter and write out the key/values from the hash intelligently as it goes.  This type of CSV format will be harder for researchers to work with, so we should ask some more questions about it.

One thought -- we should separate the act of returning the data from an experiment (the queries) from the steps required for writing that data out to a CSV.  This will make each piece easier to test and reason about, and will also make it easy for us to ensure correctness of each piece. We can still add some common code (constants, exceptions, whatever) to make sure things are tied together and all required values are returned...but lets definitely break apart the two steps into at least two objects.